src/modules/users/users.service.ts

```diff
@@ -1,28 +1,129 @@
+  ) {}
 
-  create() {
-    return 'This action adds a new user';
+  async listUsers() {
+    const users = await this.userAccountRepository.find({
+      select: ['id', 'email', 'isDeleted', 'role', 'createdAt', 'updatedAt'],
+    });
+
+    return users;
   }
 
-  findAll() {
-    return `This action returns all user`;
+  async getProfile(accountId: string) {
+    const account = await this.userAccountRepository.findOne({
+      where: { id: accountId, isDeleted: false },
+      relations: ['profile'],
+    });
+    if (!account || !account.profile) {
+      throw new NotFoundException('Account not found');
+    }
+
+    return account.profile;
+  }
+
+  async getUserWishlist(accountId: string) {
+    const account = await this.userAccountRepository.findOne({
+      where: { id: accountId, isDeleted: false },
+    });
+    if (!account) {
+      throw new NotFoundException('Account not found');
+    }
+
+    const userWishlist = await this.wishlistRepository.find({
+      where: {
+        accountId,
+      },
+    });
+
+    return userWishlist;
+  }
+
+  async updateProfile(accountId: string, updateProfileDto: UpdateProfileDto) {
+    const account = await this.userAccountRepository.findOne({
+      where: { id: accountId, isDeleted: false },
+      relations: ['profile'],
+    });
+    if (!account || !account.profile) {
+      throw new NotFoundException('Account not found');
+    }
+
+    Object.assign(account.profile, updateProfileDto);
+
+    const savedProfile = await this.userProfileRepository.save(account.profile);
+    return savedProfile;
   }
 
-  findOne() {
-    return `This action returns user`;
+  async updateProfileAvatar(accountId: string, avatarUrl: string) {
+    const account = await this.userAccountRepository.findOne({
+      where: { id: accountId, isDeleted: false },
+      relations: ['profile'],
+    });
+    if (!account || !account.profile) {
+      throw new NotFoundException('Account not found');
+    }
+
+    account.profile.image = avatarUrl;
+
+    const savedProfile = await this.userProfileRepository.save(account.profile);
+    return savedProfile;
   }
 
-  update() {
-    return `This action updates user`;
+  async toggleUserRole(accountId: string) {
+    const account = await this.userAccountRepository.findOne({
+      where: { id: accountId, isDeleted: false },
+    });
+    if (!account) {
+      throw new NotFoundException('Account not found');
+    }
+
+    account.role = account.role === Roles.Admin ? Roles.User : Roles.Admin;
+
+    const updatedUser = await this.userAccountRepository.save(account);
+    return updatedUser;
   }
 
-  remove() {
-    return `This action removes user`;
+  async deleteAccount(deleterId: string, deletingAccountId: string) {
+    const deleter = await this.userAccountRepository.findOne({
+      where: { id: deleterId, isDeleted: false },
+    });
+    if (deleter.role != Roles.Admin && deleterId != deletingAccountId) {
```

Comment:
I think the admins should not be able to delete their own account. Use OR `||` btw.

Reply:
Admin cannot delete own account? Why?

Reply:
Imagine there's only one admin account, and it'll be deleted. You should at least check the number of admins and then deny deleting if there is only one.

Reply:
FIxed